### ISSUE TITLE
[Feature Request] extend torch.arange to support 1d tensor as  parameter

### ISSUE THREAD

**Issue by zasdfgbnm** [pytorch/pytorch#23497]
## 🚀 Feature
```python
>>> torch.arange(torch.tensor([3, 4, 5]))
tensor([0, 1, 2, 0, 1, 2, 3, 0, 1, 2, 3, 4])
```

## Motivation

One possible use case is to help to deal with sequences concatenated to each other. For example:

Let's say I have a concatenated sequence, with data stored in `data`, and a long tensor `seq_id` storing which sequence the data belongs to. The `data` looks like:
```
[ --- sequence 1 ---, ----- sequence 2 -----, -- sequence 3 --]
```
and the `seq_id` looks like:
```
[0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2]
```
I want to convert this to padded sequence. Then the extended arange would help:
```
_, seq_size = torch.unique_consecutive(seq_id, return_count=True)
intra_seq_index = torch.arange(seq_size)
zero_padded_sequence = torch.zeros(seq_size.max(), seq_id.max() + 1)  # layout (sequence, batch)
zero_padded_sequence[intra_seq_index, seq_id] = data
```

**Comment by colesbury** [comment]
I'm not a fan of adding this to torch.arange, since it breaks from the NumPy API and does not follow the common broadcasting semantics.

It might make sense to add this as a stand-alone function, but I'm not sure how generally useful it would be. For things like packed sequences, I'd prefer figuring out how to handled segmented operations in general.

**Comment by eezywu** [comment]
hello, I found a solution for your question. Try:
`nums = torch.Tensor([3, 4, 5]).long()
  begin_idxes = nums.cumsum(dim=0).roll(1)
  begin_idxes[0] = 0
  result = torch.arange(nums.sum()) - begin_idxes.repeat_interleave(nums)`
